### ISSUE TITLE
PennyLane Torch Implementation Question

### ISSUE THREAD

**Issue by johnmarktaylor91** [PennyLaneAI/pennylane#5692]
Hello, I had a somewhat specific question. I maintain a library called [TorchLens](https://github.com/johnmarktaylor91/torchlens) that allows users to visualize the computational graph of PyTorch models and extract any intermediate activations. It seems to be having a hard time with PennyLane layers so I'm trying to understand why. For example, if I visualize the computational graph of a model there appears to be a gap between the input and output, suggesting that some tensor operations aren't being tracked in the middle. 

My question is--does PennyLane's PyTorch implementation only use base PyTorch operations under the hood for tensor-processing operations, or does it define new functions for performing functions on tensors? I ask because TorchLens works by temporarily decorating all functions in the PyTorch namespace such that their execution is tracked and logged, and if PennyLane performs operations on tensors that don't boil down to base PyTorch functions (e.g., that are implemented with new C code) that would explain the issue. Many thanks for any insight on this issue. 

[qnn.pdf](https://github.com/PennyLaneAI/pennylane/files/15326509/qnn.pdf)

**Comment by josh146** [comment]
Hi @johnmarktaylor91! Can I ask what PennyLane device are you using, and what differentiation method you are using?

The answer to the above depends on the underlying differentiation method --- if you are using a simulator that is written in PyTorch (such as `default.qubit`) _and_ the differentiation method is set to `backprop`, then it should be end-to-end functions on Torch tensors.

However, if you are using a non-PyTorch compatible device, or different differentiation method (such as parameter-shift or adjoint), then PennyLane is defining a custom VJP rule for the quantum device/simulator, and the quantum execution will be a black box from Torch's perspective (it could be using NumPy, C++, or even executed on real quantum hardware).

**Comment by johnmarktaylor91** [comment]
Got it, thanks so much for the fast response! I am totally new to PennyLane so I really appreciate the help. I am running it on my normal Linux computer. Here's the code I ran--based on this, it sounds like it should be all PyTorch functions under the hood, right? If so, it sounds like that is not the source of the issue and I'll keep digging ;)

```
import torchlens as tl
import torch
import pennylane as qml
from os.path import join as opj

n_qubits = 2
dev = qml.device("default.qubit", wires=n_qubits)


@qml.qnode(dev, diff_method="backprop")
def qnode(inputs, weights):
    qml.AngleEmbedding(inputs, wires=range(n_qubits))
    qml.BasicEntanglerLayers(weights, wires=range(n_qubits))
    return [qml.expval(qml.PauliZ(wires=i)) for i in range(n_qubits)]


n_layers = 6
weight_shapes = {"weights": (n_layers, n_qubits)}
qlayer = qml.qnn.TorchLayer(qnode, weight_shapes)

clayer_1 = torch.nn.Linear(2, 2)
clayer_2 = torch.nn.Linear(2, 2)
softmax = torch.nn.Softmax(dim=1)
layers = [clayer_1, qlayer, clayer_2, softmax]
model = torch.nn.Sequential(*layers)

model_history = tl.log_forward_pass(model, torch.rand(1, 2), vis_opt='unrolled',
                                    vis_outpath=opj('/home/jtaylor/projects/torchlens/local_jmt/Debugging Scripts/qnn'))
```

**Comment by johnmarktaylor91** [comment]
In case it's helpful, here is the visualization of the computational graph I'm getting (green node is input, red node is output, other nodes are tensor operations, boxes correspond to modules)--it's still showing a hole in the middle. But know that only pytorch operations are involved will help me a lot with debugging. 

[qnn.pdf](https://github.com/PennyLaneAI/pennylane/files/15327083/qnn.pdf)

**Comment by CatalinaAlbornoz** [comment]
Hi @johnmarktaylor91 ! Thank you for sharing your code. Let me investigate and get back to you in the next couple of days.

**Comment by johnmarktaylor91** [comment]
Thanks so much :)

**Comment by CatalinaAlbornoz** [comment]
Hi @johnmarktaylor91 , I'm having some trouble getting your code to run. I get an error `RecursionError: maximum recursion depth exceeded while calling a Python object`. I'll give it a try in a new environment and let you know if I can figure it out.

**Comment by johnmarktaylor91** [comment]
Thanks so much--please don't sink too much time into this, only if it's a quick fix. Much obliged :)

**Comment by CatalinaAlbornoz** [comment]
Hi @johnmarktaylor91 , I've got some good and bad news. The good news is that in a new environment I'm not getting the error that I was getting before so I can run your code with no problems. The bad news is that I don't get any output at all.

I haven't dug deep into the Torchlens documentation but do you have any pointers on what I may be doing wrong? I get no error at all but no pdf either. I'm on a mac.

**Comment by johnmarktaylor91** [comment]
Did you set `vis_outpath`? This sets the filepath for the visualization.

**Comment by CatalinaAlbornoz** [comment]
Hi @johnmarktaylor91 , I did but I had a different issue with my environment. I've now managed to fix the environment's issue so now I do get the output! I'll test it out with different PennyLane configurations and let you know what I find.

**Comment by johnmarktaylor91** [comment]
Thanks so much :)

**Comment by CatalinaAlbornoz** [comment]
Hi @johnmarktaylor91,

I've tried several PennyLane configurations and I still don't know where the black box situation occurs.

I tried even the simplest circuit and the gap that you noticed is still there. If you run the circuit below (where nothing at all is happening inside) the problem is still there. 

I'll check with the team if someone has capacity to take a look at this with fresh eyes.

```
qml.logging.enable_logging()

n_qubits = 2
dev = qml.device("default.qubit", wires=n_qubits)

@qml.qnode(dev, diff_method="backprop")
def qnode(inputs, weights):
    #print(inputs)
    #qml.RX(inputs[0][0],wires=0)
    #qml.RY(weights,wires=0)
    return [qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))] 

weight_shapes = {"weights": 1}
qlayer = qml.qnn.TorchLayer(qnode, weight_shapes)

clayer_1 = torch.nn.Linear(2, 2)
clayer_2 = torch.nn.Linear(2, 2)
softmax = torch.nn.Softmax(dim=1)
layers = [clayer_1, qlayer, clayer_2, softmax]
model = torch.nn.Sequential(*layers)
data = torch.rand(1, 2,requires_grad=False)

model_history = tl.log_forward_pass(model, data, vis_opt='unrolled',
                                    vis_outpath=...)
```

**Comment by CatalinaAlbornoz** [comment]
I think Torchlens is very cool and I'd love to see it working with PennyLane. However the answer to the mystery is escaping me at the moment. I'm hoping to get some volunteers to take a look at it in about 2 weeks. I'll keep you updated on any findings!

**Comment by johnmarktaylor91** [comment]
Thanks so much 🙏 Really appreciate all your help!

**Comment by CatalinaAlbornoz** [comment]
No problem @johnmarktaylor91 !

**Comment by kleifgch** [comment]
I decided to look into this and found a much smaller reproducer.
The ```pow```-operations at the root of the tree to output stem from:
https://github.com/PennyLaneAI/pennylane/blob/a7f26fb5b39dc9d4171cf354f8867354036ad52c/pennylane/measurements/probs.py#L206

We can verify, that this line seems to really be problematic:
```
import pennylane as qml
import torch
import torchlens as tl

class TestModule(torch.nn.Module):
    def forward(self, x):
        return qml.math.real(x) ** 2

hist = tl.log_forward_pass(
  TestModule(),
  torch.complex(torch.tensor([1.]), torch.tensor([2.])),
  vis_opt="unrolled"
)
```
is disconnected ([result.pdf](https://github.com/user-attachments/files/18722243/result_1.pdf))

Internally, `qml.math.real()` relies on autoray (specifically https://github.com/jcmgray/autoray/blob/a3147e1d2ccc3548d0338553c0a83c2da3ecbbd5/autoray/autoray.py#L1937-L1944),
which in turn accesses the attribute `real` of the tensor. Indeed, the following code results in the same graph:
```
import torch
import torchlens as tl

class TestModule(torch.nn.Module):
    def forward(self, x):
        return x.real ** 2

hist = tl.log_forward_pass(
  TestModule(),
  torch.complex(torch.tensor([1.]), torch.tensor([2.])),
  vis_opt="unrolled"
)
```

Even more extreme is this:
```
import torch
import torchlens as tl

class TestModule(torch.nn.Module):
    def forward(self, x):
        return x.real

hist = tl.log_forward_pass(
  TestModule(),
  torch.complex(torch.tensor([1.]), torch.tensor([2.])),
  vis_opt="unrolled"
)
```
which leads to `AttributeError: 'Tensor' object has no attribute 'tl_tensor_label_raw'`

@johnmarktaylor91 This should hopefully be enough for you to track down the root cause

**Comment by johnmarktaylor91** [comment]
Thanks so much for your excellent sleuthing on this, narrowing it down like this really helps the debugging. With your help I think I've managed to zoom in on the core problem, though I'm still figuring out how to fix it. The functions causing the problem rely on the `real`  and 'imag' attribute of PyTorch tensors. These attributes are in fact Python **properties** under the hood (i.e., with setter and getter functions):

<img width="852" alt="Image" src="https://github.com/user-attachments/assets/f55f3b93-2b0a-479b-ac2b-ede45fae7fcb" />

When t.real is called (where t is a tensor), it triggers the "getter" function, and creates a new tensor with the real part of t. The problem was that TorchLens isn't decorating this function when it decorates all the other functions, and I'm not sure how to apply this decorator to the getter and setter functions of the property. For all other functions, TorchLens works by simply swapping out the original functions with the decorated ones (for instance, `torch.cos = cos_decorated`). But I'm still figuring out how to do this then properties are involved. I'll keep chipping away, but if anyone knows offhand I thought I'd check :]

**Comment by kleifgch** [comment]
I see. Just for reference, the actual implementation of the property seems to be written [in C](https://github.com/pytorch/pytorch/blob/91c4bf39d39f0607833b2a226e48a9ab7262c906/torch/csrc/autograd/python_variable.cpp#L1638-L1642).
You obviously know your code better, but I think a possible solution might be to write a custom subclass of `torch.Tensor` (as described [here](https://pytorch.org/docs/stable/notes/extending.html#subclassing-torch-tensor)). Then you might be able to write:
```python
class TorchlensTensor(torch.Tensor):
  @property
  def real(self):
    tensor = super().real
    # Modify the tensor as necessary for Torchlens
    return tensor

  @real.setter
  def real(self, tensor):
    super().real = tensor
```
(Or at least something along those lines, I have not thought about the details)

For supporting pennylane in particular there is also a second solution. We could ask the autoray project to switch to `torch.real(x)` instead of `x.real` or just remove the offending code entirely since `torch.real` and `torch.imag` now also support real-valued tensors 😃

**Comment by kleifgch** [comment]
> 2. However, returning to the more elaborate model from [above](https://github.com/PennyLaneAI/pennylane/issues/5692#issuecomment-2153656555), I am still getting an interruption in the network's computational graph:

If you kept the line `#qml.RX(inputs[0][0],wires=0)` commented out as in the original comment, this is to be expected (in this case, the input actually has no bearing on the output).
If your original example is unbroken, that is a good indication that everything works fine

**Comment by johnmarktaylor91** [comment]
Good catch. If I uncomment those lines it seems to work now--this look reasonable to your eyes?

[qnn.pdf](https://github.com/user-attachments/files/18742592/qnn.pdf)

**Comment by kleifgch** [comment]
Yes, looks great! All roots of the DAG are now sensible entry points:
- `truediv_3_25:1` and `truediv_3_25:2` are there to divide the parameter of the RY gate by 2
- `astensor_1_5` belongs to the state initialization
- `astensor_19_71` and `astensor_17_56` are the eigenvalues for `_calculate_expectation()`

**Comment by johnmarktaylor91** [comment]
Amazing, sounds like mission accomplished then after this long mystery. Thanks so much for your help tracking down the bugs, definitely wouldn’t have been able to figure this out without your careful sleuthing. This was the only remaining failure case for TorchLens so I’m glad we were able to figure this out. Closing this out and will push the changes along with adding these cases to the test suite.
